cpp/src/io/parquet/decode_fixed.cu

```diff
@@ -0,0 +1,492 @@
+                     def,
+                     s->page.num_input_values);
+  }
+  __syncthreads();
+
+  // the core loop. decode batches of level stream data using rle_stream objects
+  // and pass the results to gpuUpdatePageSizes
+  int processed = 0;
+  int valid     = 0;
+  while (processed < s->page.num_input_values) {
+    int next_valid;
+
+    // only need to process definition levels if this is a nullable column
+    if (nullable) {
+      processed += def_decoder.decode_next(t);
+      __syncthreads();
+
+      next_valid =
+        gpuUpdateValidityOffsetsAndRowIndicesFlat<true, level_t>(processed, s, sb, def, t);
+    }
+    // if we wanted to split off the skip_rows/num_rows case into a separate kernel, we could skip
+    // this function call entirely since all it will ever generate is a mapping of (i -> i) for
+    // nz_idx.  gpuDecodeValues would be the only work that happens.
+    else {
+      processed += min(rolling_buf_size, s->page.num_input_values - processed);
+      next_valid =
+        gpuUpdateValidityOffsetsAndRowIndicesFlat<false, level_t>(processed, s, sb, nullptr, t);
+    }
+    __syncthreads();
+
+    // decode the values themselves
+    gpuDecodeValues(s, sb, valid, next_valid, t);
+    __syncthreads();
+
+    valid = next_valid;
+  }
+}
+
+template <typename level_t>
+__global__ void __launch_bounds__(decode_block_size) gpuDecodePageDataFixedDict(
+  PageInfo* pages, device_span<ColumnChunkDesc const> chunks, size_t min_row, size_t num_rows)
+{
+  __shared__ __align__(16) page_state_s state_g;
+  __shared__ __align__(16) page_state_buffers_s<rolling_buf_size,  // size of nz_idx buffer
+                                                rolling_buf_size,  // dictionary
+                                                1>                 // unused in this kernel
+    state_buffers;
+
+  page_state_s* const s = &state_g;
+  auto* const sb        = &state_buffers;
+  const int page_idx    = blockIdx.x;
+  const int t           = threadIdx.x;
+  PageInfo* pp          = &pages[page_idx];
+
+  if (!(BitAnd(pages[page_idx].kernel_mask, decode_kernel_mask::FIXED_WIDTH_DICT))) { return; }
+
+  // must come after the kernel mask check
+  [[maybe_unused]] null_count_back_copier _{s, t};
+
+  if (!setupLocalPageInfo(s,
+                          pp,
+                          chunks,
+                          min_row,
+                          num_rows,
+                          mask_filter{decode_kernel_mask::FIXED_WIDTH_DICT},
+                          page_processing_stage::DECODE)) {
+    return;
+  }
+
+  // the level stream decoders
+  // rolling_buf_size = 256
+  __shared__ rle_run<level_t> def_runs[rle_run_buffer_size];
+  rle_stream<level_t, decode_block_size, rolling_buf_size> def_decoder{def_runs};
+
+  // should the size be 1/2 (128?)
+  __shared__ rle_run<uint32_t> dict_runs[rle_run_buffer_size];  // should be array of 6
+  rle_stream<uint32_t, decode_block_size, rolling_buf_size> dict_stream{dict_runs};
```

Comment:
could really go crazy and parameterize this on the number of bits used for the dictionary keys.

Reply:
yes? @nvdbaranec has plans for a "grand templatization" so I think this fits right in. Lets do as follow on.